### PR TITLE
Spatial filter

### PR DESCRIPTION
**Pull request by PhilippeLafreniere18**
feat(spatialFilter / toolbar) : new spatial filter tool.

Allows the user to select address / thematics in predefined zone (SDA) or to draw his own as a polygon or circle with research radius.

### CODE REVIEW

File: packages/geo/src/lib/geometry/geometry-form-field/geometry-form-field-input.component.ts [hunk 1]
```diff
@@ -93,17 +94,46 @@ export class GeometryFormFieldInputComponent implements OnInit, OnDestroy, Contr
...
+  set drawControlIsActive(value: boolean) {
+    this._drawControlIsActive = value;
+    if (this.ready === false) {
+      return;
+    }
+    this.deactivateControl();
+    if (!this._drawControlIsActive) {
+      return;
+    } else {
+      this.createDrawControl();
```
Comment (cbourget) [inline packages/geo/src/lib/geometry/geometry-form-field/geometry-form-field-input.component.ts]: I don't think a new control should be created everytime. Maybe check if one already exists?

File: packages/geo/src/lib/geometry/geometry-form-field/geometry-form-field-input.component.ts [hunk 2]
```diff
@@ -200,6 +230,7 @@ export class GeometryFormFieldInputComponent implements OnInit, OnDestroy, Contr
     this.ready = false;
 
     this.deactivateControl();
+    this.cdRef.detach();
```
Comment (cbourget) [inline packages/geo/src/lib/geometry/geometry-form-field/geometry-form-field-input.component.ts]: Why is this line needed?

Comment (PhilippeLafreniere18) [reply packages/geo/src/lib/geometry/geometry-form-field/geometry-form-field-input.component.ts]: Had some problem at some time with destroy view and change detection, but i found an other way to resolve it, so yes, this line is no more needed. Thanks for the reminder.

File: packages/geo/src/lib/feature/shared/feature.utils.ts [hunk 3]
```diff
@@ -50,6 +50,7 @@ export function featureToOl(
   const title = getEntityTitle(feature);
   if (title !== undefined) {
     olFeature.set('_title', title, true);
+    olFeature.title = title;
```
Comment (cbourget) [inline packages/geo/src/lib/feature/shared/feature.utils.ts]: "title" isn't an attribute of ol.Feature. Why is this needed?

File: packages/geo/src/lib/feature/shared/feature.utils.ts [hunk 4]
```diff
@@ -71,6 +72,10 @@ export function featureToOl(
 
   olFeature.set('_entityRevision', getEntityRevision(feature), true);
 
+  if (feature.icon) {
+    olFeature.icon = feature.icon;
```
Comment (cbourget) [inline packages/geo/src/lib/feature/shared/feature.utils.ts]: Same comment as above

File: packages/geo/src/lib/feature/shared/feature.utils.ts [hunk 5]
```diff
@@ -109,10 +114,8 @@ export function featureFromOl(
 
   if (olLayer) {
     title = olLayer.get('title');
-    if (olLayer.get('sourceOptions')) {
-      exclude = olLayer.get('sourceOptions').excludeAttribute;
-      excludeOffline = olLayer.get('sourceOptions').excludeAttributeOffline;
-    }
+    exclude = olLayer.get('sourceOptions') ? olLayer.get('sourceOptions').excludeAttribute : undefined;
```
Comment (cbourget) [inline packages/geo/src/lib/feature/shared/feature.utils.ts]: Make a variable out of olLayer.get('sourceOptions')

**Comment by cbourget** [comment]
I don't have time to go through all of this right now but I made a few comments. It looks like an interesting feature but it's very specific. It would be great if it was a made in a way that allows different "spatial analysis services" to be connected, a bit like the search module allows different search sources.

File: packages/geo/src/lib/feature/shared/feature.utils.ts (hunk 4 above)
Comment (PhilippeLafreniere18) [reply packages/geo/src/lib/feature/shared/feature.utils.ts]: Because we need to "track" the feature icon to display it on the map when there is single point element. The icon is added to the datasource of the resulting layer.

File: packages/geo/src/lib/feature/shared/feature.utils.ts (hunk 3 above)
Comment (PhilippeLafreniere18) [reply packages/geo/src/lib/feature/shared/feature.utils.ts]: To assign a significant title at each layer added to the map (logic in the tool component).

Comment (cbourget) [reply packages/geo/src/lib/feature/shared/feature.utils.ts]: That's what properties are for. Can't you do olFeature.get('_title')

File: packages/geo/src/lib/feature/shared/feature.utils.ts (hunk 4 above)
Comment (cbourget) [reply packages/geo/src/lib/feature/shared/feature.utils.ts]: See above

File: packages/geo/src/lib/feature/shared/feature.utils.ts (hunk 3 above)
Comment (PhilippeLafreniere18) [reply packages/geo/src/lib/feature/shared/feature.utils.ts]: Seems logic, will try !

**Comment by ameliebernier** [comment]
Au lieu de "Polygone", je proposerais "Dessin". Une zone prédéfinie est aussi un polygone, donc ça montrerait mieux la différence entre les deux options.

Aussi pour "Cercle" je ne suis pas sûre... ça serait mieux "Distance" ou "Rayon" peut-être?

**Comment by ameliebernier** [comment]
Dans "Zone prédéfinie", je proposerais de trier les choix dans "Rechercher par" en ordre alphabétique

**Comment by ameliebernier** [comment]
Aussi quand on cherche une municipalité, ça se ferait de récupérer le même affichage que pour les résultats de recherche dans la barre de recherche principale? Comme ça on aurait toute l'info pertinente sur les municipalités (région administrative, dénomination...)

image

**Comment by ameliebernier** [comment]
Pour les thématiques, il ne faudrait pas en afficher qui ont un seul élément, ça alourdit l'affichage 

image

Aussi certaines thématiques me questionnent...  À quoi correspond Géographies et Secteurs?

Je classerais aussi en ordre alphabétique!

**Comment by ameliebernier** [comment]
Et finalement au lieu des onglets pour passer d'un outil à l'autre, est-ce qu'on ne devrait pas utiliser les mêmes boutons que dans l'outil de mesure... ça serait plus conséquent avec le reste de l'application

image

**Comment by PhilippeLafreniere18** [comment]
> Au lieu de "Polygone", je proposerais "Dessin". Une zone prédéfinie est aussi un polygone, donc ça montrerait mieux la différence entre les deux options.
> 
> Aussi pour "Cercle" je ne suis pas sûre... ça serait mieux "Distance" ou "Rayon" peut-être?

Atelier pour décider officiellement des labels ?

**Comment by PhilippeLafreniere18** [comment]
> Pour les thématiques, il ne faudrait pas en afficher qui ont un seul élément, ça alourdit l'affichage
> 
> image
> 
> Aussi certaines thématiques me questionnent... À quoi correspond Géographies et Secteurs?
> 
> Je classerais aussi en ordre alphabétique!

Il faudrait aussi revoir les thématiques du service !

**Comment by PhilippeLafreniere18** [comment]
> Et finalement au lieu des onglets pour passer d'un outil à l'autre, est-ce qu'on ne devrait pas utiliser les mêmes boutons que dans l'outil de mesure... ça serait plus conséquent avec le reste de l'application
> 
> image

Jumeler polygone et cercle en un onglet 'Dessin', puis deux boutons pour polygone et cercle ?

**Comment by PhilippeLafreniere18** [comment]
> Aussi quand on cherche une municipalité, ça se ferait de récupérer le même affichage que pour les résultats de recherche dans la barre de recherche principale? Comme ça on aurait toute l'info pertinente sur les municipalités (région administrative, dénomination...)
> 
> image
> 
> image

Template réutilisable comme pour le search-results ?

File: packages/geo/src/lib/geometry/geometry-form-field/geometry-form-field-input.component.ts (hunk 2 above)
Comment (PhilippeLafreniere18) [reply packages/geo/src/lib/geometry/geometry-form-field/geometry-form-field-input.component.ts]: done

File: packages/geo/src/lib/geometry/geometry-form-field/geometry-form-field-input.component.ts (hunk 1 above)
Comment (PhilippeLafreniere18) [reply packages/geo/src/lib/geometry/geometry-form-field/geometry-form-field-input.component.ts]: done

File: packages/geo/src/lib/feature/shared/feature.utils.ts (hunk 3 above)
Comment (PhilippeLafreniere18) [reply packages/geo/src/lib/feature/shared/feature.utils.ts]: done

File: packages/geo/src/lib/feature/shared/feature.utils.ts (hunk 4 above)
Comment (PhilippeLafreniere18) [reply packages/geo/src/lib/feature/shared/feature.utils.ts]: done

File: packages/geo/src/lib/feature/shared/feature.utils.ts (hunk 5 above)
Comment (PhilippeLafreniere18) [reply packages/geo/src/lib/feature/shared/feature.utils.ts]: done

File: demo/src/app/geo/spatial-filter/spatial-filter.component.ts [hunk 6]
```diff
@@ -0,0 +1,346 @@
...
+  FeatureMotion,
+  ClusterDataSource,
+  featureToOl,
+  DataSource,
+  QueryableDataSourceOptions,
+  Layer,
+} from '@igo2/geo';
+import { SpatialFilterService } from './../../../../../packages/geo/src/lib/filter/shared/spatial-filter.service';
+import { SpatialFilterQueryType } from '@igo2/geo/lib/filter/shared/spatial-filter.enum';
+import { SpatialFilterType, SpatialFilterItemType } from './../../../../../packages/geo/src/lib/filter/shared/spatial-filter.enum';
```
[the remark this answers is not in the document]
Comment (PhilippeLafreniere18) [reply demo/src/app/geo/spatial-filter/spatial-filter.component.ts]: done

File: demo/src/app/geo/spatial-filter/spatial-filter.component.ts [hunk 7]
```diff
@@ -0,0 +1,346 @@
...
+    //   results.features.meta.alias = results.features.properties.nom;
+    //   features = results.features;
+    // }
+
+    // let feature;
+    // if (features.length) {
+    //   feature = features[0];
+    // }
+    // this.selectedFeature$.next(feature);
+  }
```
[the remark this answers is not in the document]
Comment (PhilippeLafreniere18) [reply demo/src/app/geo/spatial-filter/spatial-filter.component.ts]: done

File: packages/geo/src/lib/filter/shared/spatial-filter.service.ts [hunk 8]
```diff
@@ -0,0 +1,233 @@
+import { Injectable } from '@angular/core';
+
+import { HttpClient } from '@angular/common/http';
+import { LanguageService, ConfigService } from '@igo2/core';
```
[the remark this answers is not in the document]
Comment (PhilippeLafreniere18) [reply packages/geo/src/lib/filter/shared/spatial-filter.service.ts]: done

File: packages/geo/src/lib/filter/shared/spatial-filter.service.ts [hunk 9]
```diff
@@ -0,0 +1,233 @@
...
+    'lieux.sante.etabl': 'Établissements santé',
+    'lieux.sante.gmf': 'Groupe de médecins de famille',
+    'lieux.sante.naissance': 'Naissance',
+    'lieux.sante.pharmacie': 'Pharmacie',
+    'lieux.securite.correctionnel': 'Centres correctionnels',
+    'lieux.securite.organisme': 'Organismes de sécurité',
+    'lieux.securite.palais-justice': 'Palais de justice',
+    'lieux.securite.penitencier-fed': 'Pénitenciers fédéraux',
+    'lieux.securite.penitencier-prov': 'Pénitenciers provinciaux'
+  };
```
[the remark this answers is not in the document]
Comment (PhilippeLafreniere18) [reply packages/geo/src/lib/filter/shared/spatial-filter.service.ts]: done

File: packages/geo/src/lib/filter/shared/spatial-filter.service.ts [hunk 10]
```diff
@@ -0,0 +1,233 @@
...
+          types.forEach(type => {
+            if (this.urlThematicType[type.toString()]) {
+              const language = this.languageService.getLanguage();
+              const item: SpatialFilterThematic = {
+                name: undefined
+              };
+              item.name =
+                language === 'fr' ? this.urlThematicType[type.toString()] :
+                this.languageService.translate
+                  .instant('igo.geo.spatialFilter.frenchThematics.' + this.urlThematicType[type.toString()]);
```
[the remark this answers is not in the document]
Comment (PhilippeLafreniere18) [reply packages/geo/src/lib/filter/shared/spatial-filter.service.ts]: done

File: packages/geo/src/lib/layer/shared/style.service.ts [hunk 11]
```diff
@@ -172,23 +180,25 @@ export class StyleService {
 
   createClusterStyle(feature, clusterParam: ClusterParam, layerStyle) {
     let style;
+    // const maxSize = 100;
```
[the remark this answers is not in the document]
Comment (PhilippeLafreniere18) [reply packages/geo/src/lib/layer/shared/style.service.ts]: done

File: packages/geo/src/lib/geometry/geometry-form-field/geometry-form-field-input.component.ts [hunk 12]
```diff
@@ -150,6 +209,27 @@ export class GeometryFormFieldInputComponent implements OnInit, OnDestroy, Contr
...
+    if (this.ready === false) {
+      return;
+    }
+    if (this.modifyControl.getSource()) {
+      this.modifyControl.getSource().refresh();
+    }
+    if (this.freehandDrawIsActive) {
+      let olModify;
+      setTimeout(() => {
+        olModify = this.modifyControl.olModifyInteraction;
```
Comment (cbourget) [inline packages/geo/src/lib/geometry/geometry-form-field/geometry-form-field-input.component.ts]: You could make a public method on the modify control (clearFeatures) instead of making the control public. Also, why is the setTimeout needed?

Comment (PhilippeLafreniere18) [reply packages/geo/src/lib/geometry/geometry-form-field/geometry-form-field-input.component.ts]: It seems that the control interaction is not fill before this setter. Only there to assure the modifyInteraction is not undefined before clearing the features from it.
